### PR TITLE
Add Bitcoin

### PR DESCRIPTION
**Pull request by Hargeon**
# Tasks

- [x] Deposit bitcoin
- [x] Withdraw bitcoin
- [x] Balance
- [x] Tests

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: bitcoin/bitcoin.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+func NewBitcoin(amount float64) (*Bitcoin, error) {
+	if amount < 0 {
+		return nil, InvalidPositiveAmount{amount}
+	}
+	btc := Bitcoin(amount)
+	return &btc, nil
+}
+
+// DepositBitcoin increase balance of bitcoin
+// can't add negative value
```
Comment (gordonBusyman) [inline bitcoin/bitcoin.go:16]: It is default behaviour no need to comment it

File: bitcoin/bitcoin.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+func (b *Bitcoin) DepositBitcoin(amount float64) error {
+	if amount < 0 {
+		return InvalidPositiveAmount{amount}
+	}
+	*b = *b + Bitcoin(amount)
+	return nil
+}
+
+// WithdrawBitcoin reduce balance of bitcoin
+// can't add positive value
```
Comment (gordonBusyman) [inline bitcoin/bitcoin.go:26]: Same here

File: bitcoin/bitcoin.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+func (b *Bitcoin) WithdrawBitcoin(amount float64) error {
+	if amount > 0 {
```
Comment (gordonBusyman) [inline bitcoin/bitcoin.go:28]: Why negative? 🤣

File: bitcoin/error.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package bitcoin
+
+import "fmt"
+
+type InvalidPositiveAmount struct {
```
Comment (gordonBusyman) [inline bitcoin/error.go:5]: `InvalidPositiveAmount` -> `ErrNegativeAmount`

File: bitcoin/error.go [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+	Amount float64
+}
+
+type InvalidNegativeAmount struct {
+	Amount float64
+}
```
Comment (gordonBusyman) [inline bitcoin/error.go:11]: rm

File: bitcoin/error.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+func (err InvalidPositiveAmount) Error() string {
+	return fmt.Sprintf("amount can't be negative, current amount: %v", err.Amount)
+}
+
+func (err InvalidNegativeAmount) Error() string {
+	return fmt.Sprintf("amount can't be positive, current amount: %v", err.Amount)
+}
```
Comment (gordonBusyman) [inline bitcoin/error.go:19]: Actually just:
```
var (
	// ErrNegativeAmount ...
	ErrDepositNegativeAmount = errors.New("trying to deposit negative amount") // нуили текст можно получшн сделать
	// ErrWithdrawNegativeAmount ...
	ErrWithdrawNegativeAmount = errors.New("trying to withdraw negative amount")
```
should be enough

File: wallet.go [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+type Wallet struct {
+	*bitcoin.Bitcoin
+}
+
+// NewWallet initializes wallet
+func NewWallet(b *bitcoin.Bitcoin) *Wallet {
+	wlt := Wallet{b}
+	return &wlt
+}
```
Comment (gordonBusyman) [inline wallet.go:14]: One more function can be useful - get amount of bitcoins pretty printed, like 

`w.Balance()` -> `Output: 0.0532 BTC`

**Review by gordonBusyman — CHANGES_REQUESTED** [review]

File: go.mod [hunk 8]
```diff
@@ -0,0 +1 @@
+module "github.com/Hargeon/wallet"
```
[the remark this answers is not in the document]
Comment (Hargeon) [reply go.mod:1]: Via Goland IDE(new->module)

File: wallet.go [hunk 9]
```diff
@@ -0,0 +1,14 @@
+// Package wallet holds cryptocurrency. Now supports only Bitcoin
```
[the remark this answers is not in the document]
Comment (Hargeon) [reply wallet.go:1]: > `package XXX` should be the first line of the file

Do you mean to delete the comments? [here](https://go.dev/blog/godoc) describes comment of package should be first

File: wallet.go [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+package wallet
+
+import "github.com/Hargeon/wallet/bitcoin"
+
+type Wallet struct {
+	*bitcoin.Bitcoin
+}
```
[the remark this answers is not in the document]
Comment (Hargeon) [reply wallet.go:8]: We should refactor **balance** filed if we will add new currency. I thing it is not good idea.

File: bitcoin/bitcoin.go (hunk 1 above)
Comment (Hargeon) [reply bitcoin/bitcoin.go]: Done

File: bitcoin/bitcoin.go (hunk 2 above)
Comment (Hargeon) [reply bitcoin/bitcoin.go]: Done

File: bitcoin/bitcoin.go (hunk 3 above)
Comment (Hargeon) [reply bitcoin/bitcoin.go]: Fixed)

File: bitcoin/bitcoin.go [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+		return InvalidNegativeAmount{amount}
+	}
```
[the remark this answers is not in the document]
Comment (Hargeon) [reply bitcoin/bitcoin.go:28]: Fixed

File: bitcoin/error.go (hunk 4 above)
Comment (Hargeon) [reply bitcoin/error.go]: Done

File: bitcoin/error.go (hunk 5 above)
Comment (Hargeon) [reply bitcoin/error.go]: Done

File: bitcoin/error.go (hunk 6 above)
Comment (Hargeon) [reply bitcoin/error.go]: Done

File: wallet.go (hunk 7 above)
Comment (Hargeon) [reply wallet.go:22]: Done

File: wallet.go [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+// NewWallet initializes wallet
+func NewWallet(b *bitcoin.Bitcoin) *Wallet {
+	wlt := Wallet{b}
```
[the remark this answers is not in the document]
Comment (Hargeon) [reply wallet.go]: Done

File: wallet.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (Hargeon) [reply wallet.go:1]: > Please keep the structure clean - use folders for pkgs

What do you mean? I use bitcoin folder for bitcoin package

File: bitcoin/error.go [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+
+import (
+	"errors"
+)
+
+var (
+	ErrDepositNegativeAmount  = errors.New("trying to deposit negative or zero amount")
+	ErrWithdrawNegativeAmount = errors.New("trying to withdraw negative or zero amount")
+	ErrInvalidWithdrawAmount  = errors.New("trying to withdraw more then balance")
+	ErrNewBitcoinAmount       = errors.New("trying to create bitcoin with negative amount")
```
Comment (gordonBusyman) [inline bitcoin/error.go:11]: Add comments to Exported (Public)

File: wallet.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (gordonBusyman) [reply wallet.go:1]: For the comment  you are right! Never seen is before

[the remark this answers is not in the document]
Comment (gordonBusyman) [reply wallet.go:1]: I left a comment in `wallet` package not `bitcoin`

File: go.mod (hunk 8 above)
[the remark this answers is not in the document]
Comment (gordonBusyman) [reply go.mod:1]: try to init new project via `go mod init`

File: bitcoin/bitcoin.go [hunk 14]
```diff
@@ -0,0 +1,35 @@
+// Package bitcoin helps managing bitcoin
+package bitcoin
+
+type Bitcoin float64
+
+func NewBitcoin(amount float64) (*Bitcoin, error) {
```
Comment (gordonBusyman) [inline bitcoin/bitcoin.go:6]: Add a comment it is public

File: bitcoin/bitcoin.go [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+	if amount < 0 {
+		return nil, ErrNewBitcoinAmount
+	}
+	btc := Bitcoin(amount)
+	return &btc, nil
+}
+
+func (b *Bitcoin) DepositBitcoin(amount float64) error {
```
Comment (gordonBusyman) [inline bitcoin/bitcoin.go:14]: Add a comment it is public

File: bitcoin/bitcoin.go [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+	if amount <= 0 {
+		return ErrDepositNegativeAmount
+	}
+	*b = *b + Bitcoin(amount)
+	return nil
+}
+
+func (b *Bitcoin) WithdrawBitcoin(amount float64) error {
```
Comment (gordonBusyman) [inline bitcoin/bitcoin.go:22]: Add a comment it is public

**Review by gordonBusyman — CHANGES_REQUESTED** [review]
- Please use linters, the standard set of linters will not allow the code without comments in Exported stuff
- Re-think the organization: where are you depositing your bitcoins^ in the wallet or in the bitcoin itself? (understand that bitcoin itself is a kind of wallet, but let`s move the deposit mutators to Wallet)

File: go.mod (hunk 8 above)
[the remark this answers is not in the document]
Comment (Hargeon) [reply go.mod]: Done

File: bitcoin/error.go (hunk 13 above)
Comment (Hargeon) [reply bitcoin/error.go]: Done

File: bitcoin/bitcoin.go (hunk 14 above)
Comment (Hargeon) [reply bitcoin/bitcoin.go:7]: Done

File: bitcoin/bitcoin.go (hunk 15 above)
Comment (Hargeon) [reply bitcoin/bitcoin.go]: Done

File: bitcoin/bitcoin.go (hunk 16 above)
Comment (Hargeon) [reply bitcoin/bitcoin.go]: Done

File: wallet.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (Hargeon) [reply wallet.go:1]: Done

**Review by gordonBusyman — APPROVED** [review]

**Pull request merged**